.github/workflows/release.yml

```diff
@@ -80,46 +120,90 @@ jobs:
+        path: nydus-static
+    - name: prepare release tarball
       run: |
-        HUB_VER=$(curl -s "https://api.github.com/repos/github/hub/releases/latest" | jq -r .tag_name | sed 's/^v//')
-        wget -q -O- https://github.com/github/hub/releases/download/v$HUB_VER/hub-linux-amd64-$HUB_VER.tgz | \
-        tar xz --strip-components=2 --wildcards '*/bin/hub'
-        sudo mv hub /usr/local/bin/hub
+        tag=$(echo $GITHUB_REF | cut -d/ -f3-)
+        tarball="nydus-static-$tag-${{ matrix.os }}-${{ matrix.arch }}.tgz"
+        chmod +x nydus-static/*
+        tar cf - nydus-static | gzip > ${tarball}
+        echo "tarball=${tarball}" >> $GITHUB_ENV
+    - name: store-artifacts
+      uses: actions/upload-artifact@v2
+      with:
+        name: nydus-release-tarball
+        path: ${{ env.tarball }}
+
+  # use a seperate job for darwin because github action if: condition cannot handle && properly.
+  prepare-tarball-darwin:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        arch: [amd64]
+        os: [darwin]
+    needs: [nydus-linux, nydus-macos, contrib-linux]
+    steps:
     - name: download artifacts
       uses: actions/download-artifact@v2
       with:
-        name: nydus-artifacts-${{ matrix.arch }}
+        name: nydus-artifacts-${{ matrix.os }}-${{ matrix.arch }}
         path: nydus-static
-    - name: upload artifacts
-      if: ${{ github.event_name == 'push' }}
+    - name: prepare release tarball
       run: |
         tag=$(echo $GITHUB_REF | cut -d/ -f3-)
-        tarball="nydus-static-$tag-linux-${{ matrix.arch }}.tgz"
+        tarball="nydus-static-$tag-${{ matrix.os }}-${{ matrix.arch }}.tgz"
         chmod +x nydus-static/*
         tar cf - nydus-static | gzip > ${tarball}
-        echo "tag=$tag" >> $GITHUB_ENV
-        echo "tarball=$tarball" >> $GITHUB_ENV
-    - name: Release
-      if: ${{ github.event_name == 'push' }}
+        echo "tarball=${tarball}" >> $GITHUB_ENV
+    - name: store-artifacts
+      uses: actions/upload-artifact@v2
+      with:
+        name: nydus-release-tarball
+        path: ${{ env.tarball }}
+
+  create-release:
+    runs-on: ubuntu-latest
+    needs: [prepare-tarball-linux, prepare-tarball-darwin]
+    steps:
+    - name: download artifacts
+      uses: actions/download-artifact@v2
+      with:
+        name: nydus-release-tarball
+        path: nydus-tarball
+    - name: prepare release env
+      run: |
+        cnt=0
+        for I in $(ls nydus-tarball);do cnt=$((cnt+1)); echo "tarball${cnt}=nydus-tarball/${I}" >> $GITHUB_ENV; done
+        tag=$(echo $GITHUB_REF | cut -d/ -f3-)
+        echo "tag=${tag}" >> $GITHUB_ENV
+        cat $GITHUB_ENV
+    - name: push release
+      if: github.event_name == 'push'
       uses: softprops/action-gh-release@v1
       with:
         name: "Nydus Image Service ${{ env.tag }}"
         generate_release_notes: true
-        files:
-          ${{ env.tarball }}
+        files: |
+          ${{ env.tarball1 }}
```

Comment:
Can we use an `env.tarballs` value which is split by `\n` here to simplify this?

Reply:
The newline will be eaten by $GITHUB_ENV file and we can only retrive a single line there.

Reply:
fixed